File: Silver/16173.cpp
#include <bits/stdc++.h>

using namespace std;

int n;
int MAP[4][4];
bool visited[4][4];
int dx[2] = {1, 0};
int dy[2] = {0, 1};
queue< pair<int, int> > q;

void BFS(pair<int, int> p) {
    visited[p.first][p.second] = true;
    q.push(p);

    while(!q.empty()) {
        int curX = q.front().first;
        int curY = q.front().second;
        q.pop();

        if(MAP[curX][curY] == -1) return;

        for(int j = 0; j < 2; j++) {
            int nX = curX + dx[j] * MAP[curX][curY];
            int nY = curY + dy[j] * MAP[curX][curY];

            if(nX >= 0 && nY >= 0 && nX < n && nY < n) {
                if(!visited[nX][nY]){
                    visited[nX][nY] = true;
                    q.push(make_pair(nX, nY));
                }
            }
        }
        
    }
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    cin >> n;

    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            cin >> MAP[i][j];
        }
    }

    BFS(make_pair(0, 0));

    if(!visited[n-1][n-1]) {
        cout << "Hing\n";
    } else {
        cout << "HaruHaru\n";
    }
    return 0;
}

File: Silver/16173.py
import sys
from collections import deque

input = sys.stdin.readline

n = int(input())
graph = [list(map(int, input().split())) for _ in range(n)]
visited = [[False] * n for _ in range(n)]
queue = deque()

# 이동 가능 방향은 오른쪽 or 아래
dir = [[0, 1], [1, 0]]

def BFS(x, y):
  visited[x][y] = True
  queue.append((x, y))

  while queue:
    curX, curY = queue.popleft()

    if graph[curX][curY] == -1:
      return
    
    for d in dir:
      # 현재 노드에서 오른쪽 or 아래의 점프 가능 칸 수 탐색
      nextX = curX + d[0] * graph[curX][curY]
      nextY = curY + d[1] * graph[curX][curY]

      if nextX >= 0 and nextY >= 0 and nextX < n and nextY < n:
        if visited[nextX][nextY] == False:
          visited[nextX][nextY] = True
          queue.append((nextX, nextY))
      

BFS(0, 0)

if visited[n - 1][n - 1] == False:
  print("Hing")
else:
  print("HaruHaru")